src/coins.cpp

```diff
@@ -1050,20 +1048,70 @@ bool CCoinsViewCache::HaveScRequirements(const CTransaction& tx, int height)
                 LogPrint("sc", "%s():%d - ERROR: tx [%s] tries to send funds to scId[%s] not yet created\n",
                         __func__, __LINE__, txHash.ToString(), scId.ToString() );
-                return false;
+                return state.Invalid(error("scid does not exists"),
+                     REJECT_INVALID, "sidechain-tx-scid");
             }
         }
 
         LogPrint("sc", "%s():%d - OK: tx[%s] is sending [%s] to scId[%s]\n",
             __func__, __LINE__, txHash.ToString(), FormatMoney(ft.nValue), scId.ToString());
     }
 
+    for(size_t idx = 0; idx < tx.GetVBwtRequestOut().size(); ++idx)
+    {
+        const CBwtRequestOut& mbtr = tx.GetVBwtRequestOut().at(idx);
+        const uint256& scId = mbtr.scId;
+        const boost::optional<libzendoomc::ScVk>* pWMbtrVk = nullptr;
+        if (HaveSidechain(scId))
+        {
+            if (isCeasedAtHeight(scId, height)!= CSidechain::State::ALIVE)
+            {
+                LogPrintf("ERROR: tx[%s] contains mainchain bwt request for scId[%s] already ceased at height = %d\n",
+                            txHash.ToString(), scId.ToString(), height);
+                return state.Invalid(error("received mainchain bwt request for ceased sidechain"),
+                             REJECT_INVALID, "sidechain-btr-ceased-sidechain");
+            }
+            pWMbtrVk = &(this->AccessSidechain(scId)->creationData.wMbtrVk);
+        } else
+        {
+            if (!Sidechain::hasScCreationOutput(tx, scId)) {
+                LogPrint("sc", "%s():%d - ERROR: tx [%s] contains mainchain bwt request for scId[%s] not yet created\n",
+                        __func__, __LINE__, txHash.ToString(), scId.ToString() );
+                return state.Invalid(error("scid does not exists"),
+                     REJECT_INVALID, "sidechain-tx-scid");
+            }
+            for(const auto& scCreationOut: tx.GetVscCcOut())
+            	if (scId == scCreationOut.GetScId())
+            	{
+            		pWMbtrVk = &(scCreationOut.wMbtrVk);
+            		break;
+            	}
+        }
+
+        if(!pWMbtrVk->is_initialized())
+            return state.Invalid(error("mbtr not supported"), REJECT_INVALID, "mbtr-not-supported");
+
+        // Verify mainchain bwt request proof
+        if (!scVerifier.verifyCBwtRequest(mbtr.scId, mbtr.scUtxoId, mbtr.mcDestinationAddress, mbtr.scFee, mbtr.scProof, *pWMbtrVk))
+        {
+            LogPrintf("ERROR: mbtr for scId [%s], tx[%s], pos[%d] cannot be accepted : proof verification failed\n",
+                      mbtr.scId.ToString(), tx.GetHash().ToString(), idx);
+            return state.Invalid(error("proof not verified"),
+                         REJECT_INVALID, "sidechain-mbtr-proof-not-verified");
+        }
+
+        LogPrint("sc", "%s():%d - OK: tx[%s] contains bwt transfer request for scId[%s]\n",
+            __func__, __LINE__, txHash.ToString(), scId.ToString());
+    }
+
     return true;
 }
 
```

Comment:
We need to keep track of balance changes also for BTRs (scFees). Introduce ScheduleSidechainEvent and CancelSidechainEvent for CTxMBTROut. Avoid code duplications for scheduling and canceling events for mature amounts that are also needed by FTs and SCC.

Reply:
Done in mbtr_introduction_final_fixes